src/mod/asr_tts/mod_unimrcp/mod_unimrcp.c

```diff
@@ -1962,8 +1966,8 @@ static apt_bool_t speech_on_channel_add(mrcp_application_t *application, mrcp_se
 
 	/* notify of channel open */
 	if (globals.enable_profile_events && switch_event_create_subclass(&event, SWITCH_EVENT_CUSTOM, MY_EVENT_PROFILE_OPEN) == SWITCH_STATUS_SUCCESS) {
-		orig_session = switch_core_session_locate(schannel->session_uuid);
-		if (orig_session && switch_core_session_read_lock(orig_session) == SWITCH_STATUS_SUCCESS) {
+		orig_session = switch_core_memory_pool_get_data(schannel->memory_pool, "__session");
```

Comment:
Why do we skip locking the session here? Session may be destroyed when we access the orig channel below which can cause a crash

Reply:
I removed the read read lock should in case there's an issue firing the profile_open event, so, it doesn't lock the session indefinitely. I noticed sessions were not terminating after hangup in my QA. Though, the issue was with the switch_core_session_locate method I assigned to org_session. I removed the read lock as an added precaution.

But as pointed out. I think it has more benefits adding it than removing it.

Reply:
Please use switch_core_session_locate instead of switch_core_memory_pool_get_data. If session is already in the process of destruction then switch_core_session_locate will return NULL unlike switch_core_memory_pool_get_data where you have raw access to session. If schannel somehow is destroyed then both session_uuid and memory_pool will crash regardless so I don't see any benefit of using switch_core_memory_pool_get_data

Reply:
Using `switch_core_session_locate` caused the session to lock indefinitely. I noticed this on my QA in dev. 
This is the last message when a channel is destroyed:
`switch_core_session.c:1925 Session 1 (sofia/outbound/%2B13127367736@sipdev.telnyx.com) Locked, Waiting on external entities`

Reply:
It think was happens here is that since unimrcp session is destroyed first, FS is unable to destroy the memory pool allocated to `orig_session`causing FS session to wait indefinitely. 
But using `switch_core_memory_pool_get_data` with schannel memory pool, the `orig_session` is destroy when the schannel is destroyed as well.